### ISSUE TITLE
Add a deterministic FMA

### ISSUE THREAD

**Issue by sunfishcode** [WebAssembly/relaxed-simd#44]
1. What are the instructions being proposed?

Non-quasi, guaranteed-fused FMA:

  -   f32x4.fma
  -   f32x4.fms
  -   f64x2.fma
  -   f64x2.fms

This proposal is to add these *in addition* to `qfma`, not instead of it.

2. What are the semantics of these instructions?

IEEE 754 `fusedMultiplyAdd`, and obvious subtract variant, with modifications to NaN and exception behavior as in other floating-point instructions in wasm.

3. How will these instructions be implemented? Give examples for at least
   x86-64 and ARM64. Also provide reference implementation in terms of 128-bit
   Wasm SIMD.

On x86-64 CPUs with FMA3 or FMA4, and ARM64, and other popular architectures, there is a single instruction that does this. On CPUs without an fma instruction, some options are [discussed here](https://github.com/WebAssembly/design/issues/1391).

4. How does behavior differ across processors? What new fingerprinting surfaces will be exposed?

Since wasm hides floating-point exception flags, and NaN bits are already nondeterministic, the only new differences across platforms are timings.

5. What use cases are there?

Some floating-point algorithms depend on a true fma, which is a different use case from qfma. And, some use cases want to be able to specify determinism in the wasm module, independently of whether the host implementation is enforcing determinism.

As discussed [here](https://github.com/WebAssembly/profiles/pull/1#issuecomment-935561258), it seems to make more sense to add explicit instructions for these use cases, rather than using profiles to restrict qfma to work for these use cases.

I expect one of the big questions is whether these instructions belong in relaxed-simd or should go in a separate proposal. I'm open to suggestions here. I'm starting by proposing them here, because I expect it would be confusing to users if relaxed-simd is standardized with qfma before a true fma is standardized. If a user needs a true fma, they might be tempted to use qfma if they don't (think they) care about CPUs without fma support.

**Comment by lars-t-hansen** [comment]
If these instructions were made into a proposal of their own and not tied into relaxed simd, it might make sense for a wasm implementation to provide them only on hardware that has fma natively, and let the wasm program provide its own fallback for other hardware?

**Comment by lars-t-hansen** [comment]
> I think if we want include a deterministic FMA, we also need a way for programs to check if FMA support is available in hardware.

That's why I suggested that the instruction should not be available if the hardware does not support FMA, thus allowing a feature-test of the instruction to be the test for whether the hardware supports FMA.  True, some implementations could choose to provide FMA anyway, but perhaps there's a spec-level fix for that?

**Comment by Maratyszcza** [comment]
I see a couple of issues with this proposal.

First, FMA is a not a SIMD-specific instruction, but a fundamental floating-point operation, and all processors which provide SIMD version of the FMA operation provide a scalar one too. Thus, it would be more appropriate to have FMA as a standalone proposal featuring both scalar and SIMD versions of the instructions rather than a part of Relaxed SIMD proposal, especially given that it is not processor-specific.

Secondly, I'd like to see a full version of SIMD FMA polyfill using either WAsm SIMD128 or x86 SSE4.1 intrinsics. The polyfill suggested in WebAssembly/design#1391 omits too many details, and the complete polyfills in musl and Apple LibM linked from WebAssembly/design#1391 are scalar and clearly not SIMD-friendly. AFAIK, the main application of the real FMA is for simulating higher than double-precision computations using double-double and similar representations. However, given that WAsm SIMD is limited to 128 bits and WAsm natively supports 64-bit computations, I expect that soft-float emulation of quad-precision arithmetics would be faster on pre-FMA processors than double-double SIMD operations with emulated FMA, and thus there isn't really a convincing use-case for exposing guaranteed-fused multiply-add in WAsm.

**Comment by conrad-watt** [comment]
I believe another advantage of deterministic FMA is that it may allow us to relax the way this proposal handles non-determinism to align with the existing handling of floating-point NaN (i.e. no need for the new "column" approach). WRT the [comment above](https://github.com/WebAssembly/relaxed-simd/issues/44#issuecomment-941204152), currently we need `qfma` to respect "column" non-determinism in order to faciliate `(1)`, but with `fma` and specificed-fully-non-deterministic `qfma`, (1) can be handled by `fma`, (2) can be handled by the heuristic test on `qfma` sketched by @sunfishcode just above, and (3) can be handled by `qfma`.

If possible, this would be a win for spec hygiene (no need for a new kind of non-determinism) and would address the concerns @rossberg and I raised at the phase 3 vote.

**Comment by rossberg** [comment]
I agree with @conrad-watt. Getting rid of the intrusive new form of non-determinism (a.k.a. implementation-dependent semantics) would make me much less concerned about this proposal.
